test/e2e/cpi_vm_test.go

```diff
@@ -4,46 +4,68 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"strings"
 	"time"
 
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 	"github.com/vmware/govmomi/object"
 	"github.com/vmware/govmomi/vim25/types"
 	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	k8stypes "k8s.io/apimachinery/pkg/types"
 	"k8s.io/klog/v2"
+	capvv1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
 	"sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/cluster-api/test/framework"
 	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-const machineNamespace = "default"
+const (
+	machineNamespace  = "default"
+	ControlPlaneLabel = "cluster.x-k8s.io/control-plane"
+)
 
-// getWorkerNode retrieves the worker node object for the E2E testing using workload cluster's clientset
+// getWorkerNode retrieves the first worker node object for the E2E testing using workload cluster's clientset
+// Node name should be same with VSphereVM name and only control plane VSphereVM has ControlPlaneLabel.
 func getWorkerNode() (*corev1.Node, error) {
-	nodes, err := workloadClientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	vSphereVMList := &capvv1beta1.VSphereVMList{}
```

Comment:
Might be even easier to just list Nodes in the workload cluster and exclude by the `node-role.kubernetes.io/control-plane` label.

```go
	nodes, err := workloadClientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, err
	}

	for _, node := range nodes {
		if _, ok := node.Labels["node-role.kubernetes.io/control-plane"]; ok {
			// ignore control-plane nodes
			continue
		}

		return node, nil
	}

	return nil, errors.New("worker node not found")
```

Reply:
Thanks @chrischdi! It looks neat, let me fix it and rerun tests.